lib/screens/explore_screens/vote_screens/proposals/components/vote_amount_label/interactor/usecases/get_vote_use_case.dart

```diff
@@ -2,20 +2,32 @@ import 'package:async/async.dart';
 import 'package:seeds/datasource/local/cache_repository.dart';
 import 'package:seeds/datasource/remote/api/proposals_repository.dart';
 import 'package:seeds/datasource/remote/model/vote_model.dart';
+import 'package:seeds/screens/explore_screens/vote_screens/proposals/viewmodels/proposal_view_model.dart';
 
 class GetVoteUseCase {
-  Future<Result> run(int proposalId, String account) async {
-    final cacheRepository = const CacheRepository();
-    VoteModel? voteModel = await cacheRepository.getProposalVote(account, proposalId);
-    if (voteModel == null) {
-      final result = await ProposalsRepository().getVote(proposalId, account);
+  final ProposalsRepository _proposalsRepository = ProposalsRepository();
+
+  Future<Result> run(ProposalViewModel proposal, String account) async {
+    late Result result;
+    VoteModel? voteModel;
+    if (proposal.categoryTypeLabel == 'referendum') {
+      result = await _proposalsRepository.getReferendumVote(proposal.id, account);
       if (result.isValue) {
         voteModel = result.asValue!.value as VoteModel;
-        if (voteModel.isVoted) {
-          await cacheRepository.saveProposalVote(proposalId, voteModel);
+      }
+    } else {
+      final cacheRepository = const CacheRepository();
+      voteModel = await cacheRepository.getProposalVote(account, proposal.id);
+      if (voteModel == null) {
```

Comment:
something smells about this nested logic. 

is this checking if there are proposal in the cache? 
If not, then fetch them?

Reply:
> something smells about this nested logic.

Could you be more clear?

> is this checking if there are proposal in the cache?

yes

> If not, then fetch them?

yes

Reply:
Yeah it's so complex because we can only cache if a user has voted, we can't cache the result if they have not voted yet.